var/spack/repos/builtin/packages/petsc/package.py

```diff
@@ -589,22 +578,73 @@ def test(self):
             }
             make('ex50', parallel=False)
             for feature, featureopt in testdict.items():
-                if not feature or feature in spec:
-                    self.run_test(runexe, runopt + testexe + featureopt)
-            if '+cuda' in spec:
-                make('ex7', parallel=False)
-                testexe = ['ex7', '-mat_type', 'aijcusparse',
-                           '-sub_pc_factor_mat_solver_type', 'cusparse',
-                           '-sub_ksp_type', 'preonly', '-sub_pc_type', 'ilu',
-                           '-use_gpu_aware_mpi', '0']
-                self.run_test(runexe, runopt + testexe)
-            make('clean', parallel=False)
-        w_dir = join_path(self.install_test_root, 'src/snes/tutorials')
+                if not feature or feature in self.spec:
+                    purpose_str = 'test: run ex50 example with {0}'.format(feature)
+                    self.run_test(runexe,
+                                  options=runopt + testexe + featureopt,
+                                  purpose=purpose_str,
+                                  work_dir=w_dir)
+
+    def run_ex7_test(self, runexe, runopt, w_dir):
+        """Run stand alone test: ex7"""
+
+        make('ex7', parallel=False)
+        testexe = ['ex7', '-mat_type', 'aijcusparse',
+                   '-sub_pc_factor_mat_solver_type', 'cusparse',
+                   '-sub_ksp_type', 'preonly', '-sub_pc_type', 'ilu',
+                   '-use_gpu_aware_mpi', '0']
+        purpose_str = 'test: run ex7 example with +cuda'
+        self.run_test(runexe,
+                      options=runopt + testexe,
+                      purpose=purpose_str,
+                      work_dir=w_dir)
+        make('clean', parallel=False)
+
+    def run_ex3k_test(self, runexe, runopt, w_dir):
+        """Run stand alone test: ex3k"""
+
         with working_dir(w_dir):
-            if '+kokkos' in spec:
+            if '+kokkos' in self.spec:
                 make('ex3k', parallel=False)
                 testexe = ['ex3k', '-view_initial', '-dm_vec_type', 'kokkos',
                            '-dm_mat_type', 'aijkokkos', '-use_gpu_aware_mpi', '0',
                            '-snes_monitor']
-                self.run_test(runexe, runopt + testexe)
+                purpose_str = 'run ex3k example with +kokkos'
+                self.run_test(runexe,
+                              options=runopt + testexe,
+                              purpose=purpose_str,
+                              work_dir=w_dir)
             make('clean', parallel=False)
+
```

Comment:
Given this is running in the test stage directory, `make('clean', ...)` is not needed.